RetailCoder.VBE/Navigation/CodeMetrics/CodeMetricsAnalyst.cs

```diff
@@ -0,0 +1,193 @@
+﻿using System.Collections.Generic;
+using System.Linq;
+using Rubberduck.Parsing.VBA;
+using Antlr4.Runtime.Tree;
+using Rubberduck.Parsing.Grammar;
+using Rubberduck.VBEditor;
+using Antlr4.Runtime.Misc;
+using Rubberduck.Parsing.Symbols;
+using Rubberduck.SmartIndenter;
+
+namespace Rubberduck.Navigation.CodeMetrics
+{
+    public class CodeMetricsAnalyst : ICodeMetricsAnalyst
+    {
+        private readonly IIndenterSettings _indenterSettings;
+    
+        public CodeMetricsAnalyst(IIndenterSettings indenterSettings)
+        {
+            _indenterSettings = indenterSettings;
+        }
+
+        public IEnumerable<ModuleMetricsResult> ModuleMetrics(RubberduckParserState state)
+        {
+            if (state == null || !state.AllUserDeclarations.Any())
+            {
+                // must not return Enumerable.Empty
+                yield break;
+            }
+
+            var trees = state.ParseTrees;
+
+            foreach (var moduleTree in trees)
+            {
+                yield return GetModuleResult(moduleTree.Key, moduleTree.Value, state.DeclarationFinder);
+            };
+        }
+
+        public ModuleMetricsResult GetModuleResult(RubberduckParserState state, QualifiedModuleName qmn)
+        {
+            return GetModuleResult(qmn, state.GetParseTree(qmn), state.DeclarationFinder);
+        }
+
+        private ModuleMetricsResult GetModuleResult(QualifiedModuleName qmn, IParseTree moduleTree, DeclarationFinder declarationFinder)
+        {
+            // Consider rewrite as visitor? That should make subtrees easier and allow us to expand metrics
+            var cmListener = new CodeMetricsListener(declarationFinder, _indenterSettings);
+            ParseTreeWalker.Default.Walk(cmListener, moduleTree);
+            return cmListener.GetMetricsResult(qmn);
+        }
+
+
+        private class CodeMetricsListener : VBAParserBaseListener
+        {
+            private readonly DeclarationFinder _finder;
+            private readonly IIndenterSettings _indenterSettings;
+
+            private Declaration currentMember;
+            private List<CodeMetricsResult> results = new List<CodeMetricsResult>();
+            private List<CodeMetricsResult> moduleResults = new List<CodeMetricsResult>();
+
+            private List<MemberMetricsResult> memberResults = new List<MemberMetricsResult>();
```

Comment:
Since the variables above are all instance variables, they should start with an underscore.